consensus/src/chained_bft/event_processor.rs

```diff
@@ -189,95 +179,52 @@ impl<T: Payload> EventProcessor<T> {
+            return None;
         }
         if self
             .proposer_election
-            .is_valid_proposer(proposal.proposer(), proposal.proposal.round())
+            .is_valid_proposer(proposal_msg.proposer(), proposal_msg.proposal.round())
             .is_none()
         {
             warn!(
                 "Proposer {} for block {} is not a valid proposer for this round",
-                proposal.proposer(),
-                proposal.proposal
+                proposal_msg.proposer(),
+                proposal_msg.proposal
             );
-            return ProcessProposalResult::Done(None);
+            return None;
         }
-
-        match self
-            .block_store
-            .need_fetch_for_quorum_cert(proposal.proposal.quorum_cert())
+        if let Err(e) = self
+            .sync_up(&proposal_msg.sync_info, proposal_msg.proposer())
+            .await
         {
-            NeedFetchResult::NeedFetch => {
-                return ProcessProposalResult::NeedSync(proposal);
-            }
-            NeedFetchResult::QCRoundBeforeRoot => {
-                warn!("Proposal {} has a highest quorum certificate with round older than root round {}", proposal, self.block_store.root().round());
-                return ProcessProposalResult::Done(None);
-            }
-            NeedFetchResult::QCBlockExist => {
-                if let Err(e) = self
-                    .block_store
-                    .insert_single_quorum_cert(proposal.proposal.quorum_cert().clone())
-                {
-                    warn!(
-                        "Quorum certificate for proposal {} could not be inserted to the block store: {:?}",
-                        proposal, e
-                    );
-                    return ProcessProposalResult::Done(None);
-                }
-            }
-            NeedFetchResult::QCAlreadyExist => (),
+            warn!(
+                "Dependencies of proposal {} could not be added to the block store: {:?}",
+                proposal_msg, e
+            );
+            return None;
         }
-        ProcessProposalResult::Done(self.finish_proposal_processing(proposal).await)
-    }
-
-    /// Finish proposal processing: note that multiple tasks can execute this function in parallel
-    /// so be careful with the updates. The safest thing to do is to pass the proposal further
-    /// to the proposal election.
-    /// This function is invoked when all the dependencies for the given proposal are ready.
-    async fn finish_proposal_processing(
-        &self,
-        proposal_msg: ProposalMsg<T>,
-    ) -> Option<ProposalMsg<T>> {
-        self.process_certificates(
-            proposal_msg.proposal.quorum_cert(),
-            proposal_msg.sync_info.highest_timeout_certificate(),
-        )
-        .await;
 
+        // pacemaker may catch up with the SyncInfo, check again
         let current_round = self.pacemaker.current_round();
```

Comment:
Weird, how is rust fine with the `current_round` being defined twice in the same function? (line 192)

Reply:
Yeah, it'll shadow the previous definition.